### PR TITLE
Add @_hasInitialValue to reliably mark variables with initialisers, and use it in TBDGen for StoredPropertyInitializer symbols

### PR DESCRIPTION
**Pull request by huonw**
The information about whether a variable/property is initialized is lost in the
public interface, but is, unfortunately, required because it results in a symbol
for the initializer (if a class/struct `init` is inlined, it will call
initializers for properties that it doesn't initialize itself). This is
important to preserve for TBD file generation.

Using an attribute rather than just a bit on the VarDecl means this fits into
the scheme for module interfaces: textual/valid Swift.

To validate that this is (hopefully) the only case like this, it also adds a test that is the union of all the other TBD tests, meaning multiple interesting files.

### CODE REVIEW

**Comment by huonw** [comment]
@swift-ci please smoke test

**Comment by huonw** [comment]
(Note to self: Linux failure implies need to add `// REQUIRES: objc_interop`)

I think I'll change the name of this to `@_hasInitializer`.

File: lib/AST/Decl.cpp [hunk 1]
```diff
@@ -1170,12 +1170,26 @@ SourceRange PatternBindingEntry::getOrigInitRange() const {
 }
 
 void PatternBindingEntry::setInit(Expr *E) {
+  auto PreviousInit = getInit();
   auto F = InitAndFlags.getInt();
   if (E) {
     InitAndFlags.setInt(F - Flags::Removed);
     InitAndFlags.setPointer(E);
+    if (!PreviousInit) {
```
Comment (slavapestov) [inline lib/AST/Decl.cpp]: Is there no better place for this, maybe in Sema?

File: test/IDE/print_ast_tc_decls.swift [hunk 2]
```diff
@@ -108,7 +108,7 @@ struct d0100_FooStruct {
 // PASS_COMMON-LABEL: {{^}}struct d0100_FooStruct {{{$}}
 
   var instanceVar1: Int = 0
-// PASS_COMMON-NEXT: {{^}}  var instanceVar1: Int{{$}}
+// PASS_COMMON-NEXT: {{^}}  @_bindingHasInit var instanceVar1: Int{{$}}
```
Comment (slavapestov) [inline test/IDE/print_ast_tc_decls.swift]: Can you change swift-ide-test to not print this attribute, since it will confuse users who see it in generated interfaces in Xcode?

Comment (huonw) [reply test/IDE/print_ast_tc_decls.swift]: SourceKit generally ignores it (e.g. it isn't included in the lists of attributes in its RPC protocol thing), but generated interfaces go through a different route?

File: lib/AST/Decl.cpp (hunk 1 above)
Comment (huonw) [reply lib/AST/Decl.cpp]: These initialisers get set and unset all over the place. Are you thinking that the attribute should be added once, after all of that?

File: test/IDE/print_ast_tc_decls.swift (hunk 2 above)
Comment (jrose-apple) [reply test/IDE/print_ast_tc_decls.swift]: UserInaccessible takes care of this. swift-ide-test prints all attributes by default, IIRC.

**Comment by jrose-apple** [comment]
Hm. The information is derivable from whether there are any inlinable initializers, right? Though I guess that's not just a scan of the main declaration for a struct, since they might be declared in extensions.

**Comment by huonw** [comment]
> The information is derivable from whether there are any inlinable initializers, right?

I don't think so, we'd also have to look at the body of those initialisers to see how they initialise the properties, e.g.:

```swift
struct NoSymbol {
    var x: Int
    @inlineable init() { x = 10 }
}

struct Symbol {
    var x: Int = 10
    @inlineable init() {}
}
```

Additionally, the current logic has symbol in whether or not there is an `@inlineable` initializer. Maybe we should change this?

**Comment by jrose-apple** [comment]
There are few things going on here:

- In a non-resilient library, the property initializers always have to be public, because until [SE-0189](https://github.com/apple/swift-evolution/blob/master/proposals/0189-restrict-cross-module-struct-initializers.md) anyone could add their own non-delegating initializer. It's unclear whether we'll be able to break that in later releases given that it was supported in Swift 4.0 and thus only produces a warning under `-swift-version 4` and `-swift-version 4.2`.

- With a resilient library, SE-0189 is always enforced, so you only need to worry about initializers within the module. You obviously don't need to worry about non-inlinable initializers, which includes all non-fixed-contents structs. So the choice of whether or not to give the initial values public symbols is controlled by whether the struct is fixed-contents; it could check for inlinable initializers instead. (@slavapestov, I may have gotten some details of this wrong. Does it make the initial value expressions public or shared-linkage serialized?)

**Comment by huonw** [comment]
> In a non-resilient library, the property initializers always have to be public, because until SE-0189 anyone could add their own non-delegating initializer. It's unclear whether we'll be able to break that in later releases given that it was supported in Swift 4.0 and thus only produces a warning under -swift-version 4 and -swift-version 4.2.

Ah, okay. It seems this case contradicts your previous point, as it won't be covered by scanning a module for `@inlineable init`. Additionally, even in an always-SE-0189 world, we won't want to be emitting public symbols for variables that don't have public initializers meaning it seems like we need a patch like this one, in some form.

I think I might be missing something?

File: lib/AST/Decl.cpp (hunk 1 above)
Comment (huonw) [reply lib/AST/Decl.cpp]: It does seem like there's a lot of variables created and not in several places throughout the compiler and REPL, etc., and it seems not all of them go through the "obvious" type checker paths (e.g. property behavior's `completePropertyBehaviorStorage` sets everything up perfectly, and skips the main checker). It seems unfortunate to push this detail to be repeated at all such code paths (possibly including in lldb), given it is entirely computable from info given to this call, but I agree that it's a little ugly to have it here.

**Comment by jrose-apple** [comment]
Now that I think about it, I suspect we *don't* want to add public symbols for initial value expressions in a resilient build at all. That would mean adding or removing one would be a breaking change. That in turn suggests serializing them along with the initializer.

In the non-resilient case, I think you're right, I got it wrong the first time. We always need these symbols.

**Comment by huonw** [comment]
> That in turn suggests serializing them along with the initializer.

This means we need to do a `@usableFromInlineable` check for everything the initialiser uses. In a (hypothetical) world where everything is explicit, it seems we should even require people to mark the initialiser itself `@inlinable`/`@usableFromInlineable`, since changes to it won't be reflected downstream?

**Comment by jrose-apple** [comment]
That's also a little scary, since it means that a `let` can have a different value from a cross-module initializer if they ignore the warning in Swift 4 mode. But I guess "you were warned".

**Comment by slavapestov** [comment]
@jrose-apple Nice, I didn't even realize that thing about the 'let'. Good thing we banned this, then. :) Note that it *does* do the right thing in the cross-file case in non-WMO mode, and it doesn't even have to re-typecheck the expression unless its also used to infer the type.

**Comment by jrose-apple** [comment]
I still don't think we need an attribute at all. It sounds like it's fully derivable from the description we have above. None of that requires looking at initializer bodies.

(And yes, I'm also in favor of "initial value" since "initializer" is taken.)

**Comment by huonw** [comment]
I don't understand what you mean by "fully derivable"? The problem this patch is solving is when generating a TBD file by merging modules, or by looking at the textual interface, there's absolutely no record of an initial value existing. I assume we don't want to have symbols for an initialiser for every property "just in case", that is, we only want symbols for ones with actual initial values.

**Comment by huonw** [comment]
@swift-ci please smoke test

File: lib/AST/Decl.cpp (hunk 1 above)
Comment (huonw) [reply lib/AST/Decl.cpp]: I've moved it to `TypeChecker::typeCheckPatternBinding`.

File: include/swift/AST/Attr.def [hunk 3]
```diff
@@ -376,6 +376,10 @@ SIMPLE_DECL_ATTR(_forbidSerializingReference, ForbidSerializingReference,
   OnAnyDecl |
   LongAttribute | RejectByParser | UserInaccessible | NotSerialized,
   77)
+SIMPLE_DECL_ATTR(_hasInitialValue, HasInitialValue,
+  OnVar |
+  SILOnly | UserInaccessible,
```
Comment (jrose-apple) [inline include/swift/AST/Attr.def]: Oops, I think `UserInaccessible` is fine here. No need to limit it to SIL mode; no one's actually going to try to type it.

File: lib/AST/ASTVerifier.cpp [hunk 4]
```diff
@@ -250,6 +250,16 @@ class Verifier : public ASTWalker {
     GenericEnv.push_back({DC});
   }
 
+  /// \brief Returns the SourceFile for this AST, if it was parsed from a .swift
+  /// or (if \c allowSIL) .sil file.
```
Comment (jrose-apple) [inline lib/AST/ASTVerifier.cpp]: Why this change? That is, what condition doesn't hold for this new attribute?

File: lib/Sema/CodeSynthesis.cpp [hunk 5]
```diff
@@ -1324,7 +1325,7 @@ void TypeChecker::completePropertyBehaviorStorage(VarDecl *VD,
   Storage->setImplicit();
   Storage->setAccess(AccessLevel::Private);
   Storage->setSetterAccess(AccessLevel::Private);
-  
+
```
Comment (jrose-apple) [inline lib/Sema/CodeSynthesis.cpp]: Sorry, this file only has whitespace changes now; can you remove them?

File: include/swift/AST/Decl.h [hunk 6]
```diff
@@ -1862,8 +1862,11 @@ class PatternBindingEntry {
 public:
   PatternBindingEntry(Pattern *P, SourceLoc EqualLoc, Expr *E,
                       DeclContext *InitContext)
-      : ThePattern(P), EqualLoc(EqualLoc), InitAndFlags(E, {}),
-        InitContext(InitContext) {}
+      : ThePattern(P), EqualLoc(EqualLoc), InitAndFlags(nullptr, {}),
+        InitContext(InitContext) {
+    // Setting the initializer does some book-keeping, so we do that separately.
+    setInit(E);
```
Comment (jrose-apple) [inline include/swift/AST/Decl.h]: This doesn't seem to be true anymore. Maybe it can go back to the way it was?

File: lib/TBDGen/TBDGen.cpp [hunk 7]
```diff
@@ -43,16 +43,17 @@ static bool isGlobalOrStaticVar(VarDecl *VD) {
...
-      auto declRef =
-          SILDeclRef(var, SILDeclRef::Kind::StoredPropertyInitializer);
-      // Stored property initializers for public properties are currently
-      // public.
-      addSymbol(declRef);
-    }
+    entry.getPattern()->forEachVariable([&](VarDecl *VD) {
+      // Non-global variables might have an explicit initializer symbol.
+      if (VD->getAttrs().hasAttribute<HasInitialValueAttr>() &&
+          !isGlobalOrStaticVar(VD)) {
```
Comment (jrose-apple) [inline lib/TBDGen/TBDGen.cpp]: Oh no, I forgot that this is also relevant for classes. \*sigh\* Seems reasonable. Is it worth checking to see if this is a resilient module, though? (ModuleDecl::getResilienceStrategy)

**Review by jrose-apple — COMMENTED** [review]
Okay, I think this looks pretty good, then, at least for non-resilient libraries. Sorry for misunderstanding the problem and going off in the wrong direction.

File: include/swift/AST/Attr.def (hunk 3 above)
Comment (huonw) [reply include/swift/AST/Attr.def]: 👍

File: include/swift/AST/Decl.h (hunk 6 above)
Comment (huonw) [reply include/swift/AST/Decl.h]: Ah, yeah. This and the change in ASTVerifier and CodeSynthesis are all just left overs from previous versions. Reverted.

File: lib/TBDGen/TBDGen.cpp (hunk 7 above)
Comment (huonw) [reply lib/TBDGen/TBDGen.cpp]: Yes, might as well do that change here.

**Comment by huonw** [comment]
@swift-ci please smoke test

**Comment by huonw** [comment]
@swift-ci please smoke test

**Comment by slavapestov** [comment]
LGTM
